descartes_light/solvers/src/bgl/bgl_solver.cpp

```diff
@@ -19,31 +19,53 @@
 #include <descartes_light/solvers/bgl/impl/bgl_dijkstra_solver.hpp>
 #include <descartes_light/solvers/bgl/impl/utils.hpp>
 
+#include <descartes_light/descartes_macros.h>
+DESCARTES_IGNORE_WARNINGS_PUSH
+#include <boost/graph/visitors.hpp>
+#include <boost/preprocessor.hpp>
+DESCARTES_IGNORE_WARNINGS_POP
+
+// Macro for explicitly instantiating a class with a variable number of template arguments
+#define INSTANTIATE(T, ...) template class T<__VA_ARGS__>;
+
+// Implementation macro for explicitly instantiating a class for a specific element of the Cartesian product
+// (class_name, (template params))
+#define INSTANTIATE_PRODUCT_IMPL(r, product)                                                                           \
+  INSTANTIATE(BOOST_PP_SEQ_HEAD(product), BOOST_PP_SEQ_ENUM(BOOST_PP_SEQ_TAIL(product)))
+
+/**
+ * Macro for instantiating a template class for a Cartesian product of float types and event visitor types
+ * For example, instantiation of class Foo with types ((float)(double)) and visitors ((A)(B)) would result in 4
+ * instantiations:
+ *   - Foo<float, A>
+ *   - Foo<float, B>
+ *   - Foo<double, A>
+ *   - Foo<double, B>
+ */
+#define INSTANTIATE_PRODUCT(TEMPLATE, FLOAT_TYPES, EVENT_VISITORS)                                                     \
+  BOOST_PP_SEQ_FOR_EACH_PRODUCT(INSTANTIATE_PRODUCT_IMPL, ((TEMPLATE))(FLOAT_TYPES)(EVENT_VISITORS))
+
 namespace descartes_light
 {
-// Explicit template instantiation
-// Partial implementations
-template class BGLSolverBase<double>;
-template class BGLSolverBase<float>;
-
-template class BGLSolverBaseSVDE<double>;
-template class BGLSolverBaseSVDE<float>;
+// Event visitors
+template struct early_terminator<boost::on_examine_vertex>;
 
-template class BGLSolverBaseSVSE<double>;
-template class BGLSolverBaseSVSE<float>;
+template struct add_all_edges_dynamically<float, boost::on_examine_vertex>;
+template struct add_all_edges_dynamically<double, boost::on_examine_vertex>;
 
-// Full Implementations
-template class BGLDijkstraSVSESolver<double>;
-template class BGLDijkstraSVSESolver<float>;
+// Explicit template instantiation
+#define FLOAT_TYPES (double)(float)
+#define DIJKSTRA_EVENT_VISITORS (boost::null_visitor)(early_terminator<boost::on_examine_vertex>)
```

Comment:
Would this be the standard for defining new families of solvers?

Reply:
Yes, I think so, just because the event visitors will likely be different between search algorithms. The DFS solver could use an early terminator like Dijkstra, but the event filter would have to be `boost::on_discover_vertex` instead

Reply:
[This](https://github.com/colin-lewis-19/descartes_light/blob/1ffb1a32aa1b6f727f97844185bc8291e1948790/descartes_light/solvers/src/bgl/bgl_solver.cpp#L60) is what I did in #77 for DFS